Add matrix halo exchange to "binning" interpolation method.

odlomax opened the pull request:
### Description

This PR addresses an error in the "binning" interpolation method (really need to come up with a better name...), introduced by myself and @mo-lormi  a while back.

In summary, the original method transposes a "low-to-high-resolution" interpolation matrix to create a "high-to-low-resolution" projection, notionally smoothing or "binning" the high resolution data.

The initial implementation simply transposes the interpolation matrix and renormalises each row of the transposes to sum to 1. So far so good.

However, this introduces some problems, as we now have matrix rows which sometimes project into ghost elements of fields. Not only this, if some of the source field is projected into into a ghost element, then that source information *isn't* projected into the corresponding owned element. This error was largely hidden, as the local weight normalisation followed by a halo exchange makes it difficult to spot.

We can see the problem if we look at some "binning" kernels from a CS-LFR-48 -> CS-LFR-12 interpolation matrix, shown below:

<img width="736" height="772" alt="kernels_no_halo_exchange" src="https://github.com/user-attachments/assets/d0fe819b-90c3-42b5-97b6-0983452ab38f" />


Here you see that some of the kernel is _chopped off_ at the partition boundaries around each face of the cube.


The solution to the problem is the perform a halo exchange (of sorts) on the transposed matrix. Here, any row that corresponds to a ghost element is sent to the owning PE and incorporated into its interpolation matrix. After doing this, the kernels are no longer as distorted:

<img width="736" height="772" alt="kernels" src="https://github.com/user-attachments/assets/d49d980b-0607-4ae3-966b-cf3d80bb7d9f" />


### Contributor Declaration

By opening this pull request, I affirm the following:

* All authors agree to the [Contributor License Agreement](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md).
* The code follows the project's coding standards.
* I have performed self-review and added comments where needed.
* I have added or updated tests to verify that my changes are effective and functional.
* I have run all existing tests and confirmed they pass.

pmaciel commented:
Hi @odlomax , nice to see you around after quite some time! I've written a very similar method a few years ago, doing exactly this - transposing and re-normalising. But, it had exactly the same problems that you mention. I was using it for computationaly "heavy" methods, like the grid-box average conservative method, to build the matrix the cheapest way then transposing.

When I tested it, the results had these problems -- they come from the fact that the original weights reflect the element-related contribution to each node, normalised at that specific target point. This information is not transposable, without rebalancing the element-related information -- that is, before re-normalising the weights need to be re-scaled with the original supporting elements information (in my case, with their area). The way I foudn about this, is to re-caclulate the grid-box area for the input/output then rebalance (more costly than transpose + renormalise), and in this case I end up with the correct (exactly so) result. This might be your situation (I haven't looked in detail). I didn't implement finally it because there was no use case at the time (the GBA method is affordable).

Just sharing some experience :-) 👍

wdeconinck reviewed (changes requested):
Hi @odlomax this looks quite good! Just a few suggestions.

wdeconinck commented on src/atlas/interpolation/method/binning/Binning.cc:
```diff
@@ -5,193 +5,262 @@
-  smx_interp_tr.transpose(); // transpose the copy in-place
+    const auto interp              = Interpolation(interpAncillaryScheme_, target_, source_);
+    const auto interpMatrixStorage = interpolation::MatrixCache(interp).matrix();
+    auto interpMatrix              = linalg::make_non_owning_eckit_sparse_matrix(interpMatrixStorage);
```
You can make this more general by only using View and Storage:
```
SparseMatrixView<Scalar,Index> interpMatrixView = atlas::linalg::make_host_view<Scalar,Index>(interpMatrixStorage);
```
and refactor the transpose+halo-exchange routine to create a new MatrixStorage from the view:
```c++
SparseMatrixStorage Binning::transposeAndHaloExchange(SparseMatrixView<Scalar,Index> interpMatrix) const;
```
In this function you use an iterator which is not implemented as of yet in `SparseMatrixView`, so a regular (also more efficient) for loop would be required then, as here: https://github.com/ecmwf/atlas/blob/develop/src/atlas/linalg/sparse/SparseMatrixToTriplets.h#L24

wdeconinck reviewed (changes requested):
some more suggestions

odlomax commented on src/atlas/interpolation/method/binning/Binning.cc:
```diff
@@ -5,193 +5,262 @@
+    mpi::comm().allToAll(sendBuffer, recvBuffer);
+
+    // Create a map from source partition and remote index to local index for ghost points only.
+    auto sourceRemoteToLocalMap = std::map<std::pair<int, idx_t>, idx_t>{};
```
To my shame, I hadn't realised that std::map uses a comparison operator instead of a hash. I'm surprised this worked at all!

odlomax commented:
I've incorrectly merged/rebased the branch. I'll update you when it's fixed!

odlomax commented:
Okie dokie, I think it's fixed. I messed up a rebase, so I had to nuke the branch and rebuild it from develop.

- I've added `vector::reserve` when I know the size of the array.
- In classic over-engineering style, I've added a small `linalg::Triplet` class (oddly missing in `SparseMatrixToTriplets.h`), with to-sparse-matrix and for-each methods. This completely removes the dependency on `eckit::linalg::SparseMatrix`.
- I've removed the hash specialisation. It wasn't being used, as you picked up on. Also I'm not entirely sure if it's UB or not. `std::map<std::pair<int, int>, int>` works just fine.
- I've fixed the typos you found.

Cheers!

wdeconinck commented on src/atlas/linalg/sparse/SparseMatrixTriplet.h:
```diff
@@ -0,0 +1,100 @@
+    Value value_{};
+};
+
+template <typename Value, typename Index>
+SparseMatrixStorage make_sparse_matrix_storage_from_triplets(Index n_rows, Index n_cols,
+                                                             std::vector<Triplet<Value, Index>>&& triplets) {
```
Internally the vector is not moved anywhere and gets compressed (copied) into the sparse matrix storage arrays, so it's not a matter of transferring (moving) ownership. Hence I prefer the lvalue, but an rvalue version can live besides it if you want to be support inline creation of the triplet vector. With above delegation approach it's essentially a one-liner so to speak.
